core/shared/src/main/scala/cats/parse/Parser.scala

```diff
@@ -1599,6 +1607,23 @@ object Parser extends ParserInstances {
       override def parseMut(state: State): B = Impl.map(parser, fn, state)
     }
 
+    final def branch[A, B, C](
+        parser: Parser[Either[A, B]],
+        l: Parser[A => C],
+        r: Parser[B => C],
+        state: State
+    ): C = {
+      val either = parser.parseMut(state)
+      if ((state.error eq null) && state.capture)
+        either.fold(a => l.ap(Parser.pure(a)), b => r.ap(Parser.pure(b))).parseMut(state)
```

Comment:
Let's not use fold. Pattern matching avoids allocating a closure and is generally faster, and this could appear in a tight loop inside a rep.

Reply:
:+1: I'll inline the `valueOr` that replaced it.

Reply:
Cool. And thinking more, I don't think we want to use ap. We should run parseMut on the currect side then apply the function ourselves.

Reply:
Yep, already rid of that locally. I was translating a bit too literally from Haskell last night.